core/src/Kokkos_ExecPolicy.hpp

```diff
@@ -589,6 +589,14 @@ class TeamPolicy
   TeamPolicy(const internal_policy& p) : internal_policy(p) {
     first_arg = false;
   }
+  void check_valid_storage_level_argument(int level) const {
+    if (!(level == 0 || level == 1)) {
+      std::stringstream ss;
+      ss << "TeamPolicy::set_scratch_size(/*level*/ " << level
+         << ", ...) storage level argument must be 1 or 2 to be valid\n";
+      Impl::throw_runtime_exception(ss.str());
```

Comment:
Suggestion: move this to a function named `throw_invalid_storage_exception(level)` or something like that and move it's implementation to `Kokkos_ExecPolicy.hpp`. Then remove `#include <sstream>` from this header file (it's just a start, but we have to start somewhere. Throwing invalid scratch size policy exceptions isn't performance-sensitive enough that it needs to be inlined.)

Reply:
(IMHO, the `if` statement itself should stay in a header file, since inlining it doesn't require any additional header files and it's used far more than the body of the if statement)

Reply:
I turned it into a free function and move the implementation to `impl/Kokkos_ExecPolicy.cpp`.
I decided to keep the if-statement because it did not feel right to separate things after I realized the error message said "must be 1 or 2"  instead of "0 or 1"